{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting scvi\n",
      "  Using cached scvi-0.6.8-py2.py3-none-any.whl.metadata (12 kB)\n",
      "Collecting scib\n",
      "  Using cached scib-1.1.5-1-py3-none-any.whl.metadata (9.8 kB)\n",
      "Collecting bbknn\n",
      "  Using cached bbknn-1.6.0-py3-none-any.whl.metadata (8.2 kB)\n",
      "Requirement already satisfied: numpy>=1.16.2 in /opt/conda/lib/python3.10/site-packages (from scvi) (1.26.4)\n",
      "Requirement already satisfied: torch>=1.3 in /opt/conda/lib/python3.10/site-packages (from scvi) (2.0.0.post104)\n",
      "Requirement already satisfied: matplotlib>=3.1.2 in /opt/conda/lib/python3.10/site-packages (from scvi) (3.9.1)\n",
      "Requirement already satisfied: scikit-learn>=0.20.3 in /opt/conda/lib/python3.10/site-packages (from scvi) (1.4.2)\n",
      "Requirement already satisfied: h5py>=2.9.0 in /opt/conda/lib/python3.10/site-packages (from scvi) (3.11.0)\n",
      "Requirement already satisfied: pandas>=0.25 in /opt/conda/lib/python3.10/site-packages (from scvi) (2.1.4)\n",
      "Requirement already satisfied: tqdm>=4.31.1 in /opt/conda/lib/python3.10/site-packages (from scvi) (4.66.5)\n",
      "Collecting xlrd>=1.2.0 (from scvi)\n",
      "  Using cached xlrd-2.0.1-py2.py3-none-any.whl.metadata (3.4 kB)\n",
      "Collecting hyperopt==0.1.2 (from scvi)\n",
      "  Using cached hyperopt-0.1.2-py3-none-any.whl.metadata (1.2 kB)\n",
      "Requirement already satisfied: anndata>=0.7 in /opt/conda/lib/python3.10/site-packages (from scvi) (0.10.9)\n",
      "Requirement already satisfied: scanpy>=1.4.6 in /opt/conda/lib/python3.10/site-packages (from scvi) (1.10.2)\n",
      "Collecting scikit-misc>=0.1.3 (from scvi)\n",
      "  Using cached scikit_misc-0.5.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.0 kB)\n",
      "Requirement already satisfied: scipy in /opt/conda/lib/python3.10/site-packages (from hyperopt==0.1.2->scvi) (1.11.4)\n",
      "Collecting pymongo (from hyperopt==0.1.2->scvi)\n",
      "  Using cached pymongo-4.8.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (22 kB)\n",
      "Requirement already satisfied: six in /opt/conda/lib/python3.10/site-packages (from hyperopt==0.1.2->scvi) (1.16.0)\n",
      "Requirement already satisfied: networkx in /opt/conda/lib/python3.10/site-packages (from hyperopt==0.1.2->scvi) (3.3)\n",
      "Requirement already satisfied: future in /opt/conda/lib/python3.10/site-packages (from hyperopt==0.1.2->scvi) (1.0.0)\n",
      "Requirement already satisfied: seaborn in /opt/conda/lib/python3.10/site-packages (from scib) (0.13.2)\n",
      "Requirement already satisfied: numba in /opt/conda/lib/python3.10/site-packages (from scib) (0.60.0)\n",
      "Requirement already satisfied: leidenalg in /opt/conda/lib/python3.10/site-packages (from scib) (0.10.2)\n",
      "Requirement already satisfied: umap-learn in /opt/conda/lib/python3.10/site-packages (from scib) (0.5.6)\n",
      "Collecting pydot (from scib)\n",
      "  Using cached pydot-3.0.1-py3-none-any.whl.metadata (9.9 kB)\n",
      "Requirement already satisfied: igraph>=0.10 in /opt/conda/lib/python3.10/site-packages (from scib) (0.11.6)\n",
      "Requirement already satisfied: llvmlite in /opt/conda/lib/python3.10/site-packages (from scib) (0.43.0)\n",
      "Collecting deprecated (from scib)\n",
      "  Using cached Deprecated-1.2.14-py2.py3-none-any.whl.metadata (5.4 kB)\n",
      "Collecting Cython (from bbknn)\n",
      "  Using cached Cython-3.0.11-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.2 kB)\n",
      "Collecting annoy (from bbknn)\n",
      "  Using cached annoy-1.17.3-cp310-cp310-linux_x86_64.whl\n",
      "Requirement already satisfied: pynndescent in /opt/conda/lib/python3.10/site-packages (from bbknn) (0.5.13)\n",
      "Requirement already satisfied: array-api-compat!=1.5,>1.4 in /opt/conda/lib/python3.10/site-packages (from anndata>=0.7->scvi) (1.8)\n",
      "Requirement already satisfied: exceptiongroup in /opt/conda/lib/python3.10/site-packages (from anndata>=0.7->scvi) (1.2.2)\n",
      "Requirement already satisfied: natsort in /opt/conda/lib/python3.10/site-packages (from anndata>=0.7->scvi) (8.4.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /opt/conda/lib/python3.10/site-packages (from anndata>=0.7->scvi) (24.1)\n",
      "Requirement already satisfied: texttable>=1.6.2 in /opt/conda/lib/python3.10/site-packages (from igraph>=0.10->scib) (1.7.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /opt/conda/lib/python3.10/site-packages (from matplotlib>=3.1.2->scvi) (1.2.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /opt/conda/lib/python3.10/site-packages (from matplotlib>=3.1.2->scvi) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /opt/conda/lib/python3.10/site-packages (from matplotlib>=3.1.2->scvi) (4.53.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /opt/conda/lib/python3.10/site-packages (from matplotlib>=3.1.2->scvi) (1.4.5)\n",
      "Requirement already satisfied: pillow>=8 in /opt/conda/lib/python3.10/site-packages (from matplotlib>=3.1.2->scvi) (10.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /opt/conda/lib/python3.10/site-packages (from matplotlib>=3.1.2->scvi) (3.1.2)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /opt/conda/lib/python3.10/site-packages (from matplotlib>=3.1.2->scvi) (2.9.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /opt/conda/lib/python3.10/site-packages (from pandas>=0.25->scvi) (2023.3)\n",
      "Requirement already satisfied: tzdata>=2022.1 in /opt/conda/lib/python3.10/site-packages (from pandas>=0.25->scvi) (2024.1)\n",
      "Requirement already satisfied: joblib in /opt/conda/lib/python3.10/site-packages (from scanpy>=1.4.6->scvi) (1.4.2)\n",
      "Requirement already satisfied: legacy-api-wrap>=1.4 in /opt/conda/lib/python3.10/site-packages (from scanpy>=1.4.6->scvi) (1.4)\n",
      "Requirement already satisfied: patsy in /opt/conda/lib/python3.10/site-packages (from scanpy>=1.4.6->scvi) (0.5.6)\n",
      "Requirement already satisfied: session-info in /opt/conda/lib/python3.10/site-packages (from scanpy>=1.4.6->scvi) (1.0.0)\n",
      "Requirement already satisfied: statsmodels>=0.13 in /opt/conda/lib/python3.10/site-packages (from scanpy>=1.4.6->scvi) (0.14.2)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /opt/conda/lib/python3.10/site-packages (from scikit-learn>=0.20.3->scvi) (3.5.0)\n",
      "Requirement already satisfied: filelock in /opt/conda/lib/python3.10/site-packages (from torch>=1.3->scvi) (3.15.4)\n",
      "Requirement already satisfied: typing-extensions in /opt/conda/lib/python3.10/site-packages (from torch>=1.3->scvi) (4.12.2)\n",
      "Requirement already satisfied: sympy in /opt/conda/lib/python3.10/site-packages (from torch>=1.3->scvi) (1.13.0)\n",
      "Requirement already satisfied: jinja2 in /opt/conda/lib/python3.10/site-packages (from torch>=1.3->scvi) (3.1.4)\n",
      "Requirement already satisfied: wrapt<2,>=1.10 in /opt/conda/lib/python3.10/site-packages (from deprecated->scib) (1.14.1)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /opt/conda/lib/python3.10/site-packages (from jinja2->torch>=1.3->scvi) (2.1.5)\n",
      "Collecting dnspython<3.0.0,>=1.16.0 (from pymongo->hyperopt==0.1.2->scvi)\n",
      "  Using cached dnspython-2.6.1-py3-none-any.whl.metadata (5.8 kB)\n",
      "Requirement already satisfied: stdlib-list in /opt/conda/lib/python3.10/site-packages (from session-info->scanpy>=1.4.6->scvi) (0.10.0)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /opt/conda/lib/python3.10/site-packages (from sympy->torch>=1.3->scvi) (1.3.0)\n",
      "Using cached scvi-0.6.8-py2.py3-none-any.whl (158 kB)\n",
      "Using cached hyperopt-0.1.2-py3-none-any.whl (115 kB)\n",
      "Using cached scib-1.1.5-1-py3-none-any.whl (79 kB)\n",
      "Using cached bbknn-1.6.0-py3-none-any.whl (14 kB)\n",
      "Using cached scikit_misc-0.5.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (188 kB)\n",
      "Using cached xlrd-2.0.1-py2.py3-none-any.whl (96 kB)\n",
      "Using cached Cython-3.0.11-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.6 MB)\n",
      "Using cached Deprecated-1.2.14-py2.py3-none-any.whl (9.6 kB)\n",
      "Using cached pydot-3.0.1-py3-none-any.whl (22 kB)\n",
      "Using cached pymongo-4.8.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.2 MB)\n",
      "Using cached dnspython-2.6.1-py3-none-any.whl (307 kB)\n",
      "Installing collected packages: annoy, xlrd, scikit-misc, pydot, dnspython, deprecated, Cython, pymongo, hyperopt, bbknn, scvi, scib\n",
      "Successfully installed Cython-3.0.11 annoy-1.17.3 bbknn-1.6.0 deprecated-1.2.14 dnspython-2.6.1 hyperopt-0.1.2 pydot-3.0.1 pymongo-4.8.0 scib-1.1.5 scikit-misc-0.5.1 scvi-0.6.8 xlrd-2.0.1\n"
     ]
    }
   ],
   "source": [
    "!pip install scvi scib bbknn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/scvi/__init__.py:31: DeprecationWarning: scvi is deprecated, please uninstall scvi via `pip uninstall scvi` and install the new scvi-tools package at github.com/YosefLab/scvi-tools\n",
      "  warnings.warn(deprecation_msg, DeprecationWarning)\n",
      "2024-09-10 10:55:13.348472: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "/opt/conda/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:368: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/opt/conda/lib/python3.10/site-packages/rpy2/robjects/numpy2ri.py:252: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/tmp/ipykernel_4495/1582476536.py:18: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  anndata2ri.activate()\n"
     ]
    }
   ],
   "source": [
    "# Python packages\n",
    "import scanpy as sc\n",
    "import scvi\n",
    "import bbknn\n",
    "import scib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "\n",
    "# R interface\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects import r\n",
    "import rpy2.rinterface_lib.callbacks\n",
    "import anndata2ri\n",
    "\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:990: DeprecationWarning: The `source` parameter emit a  deprecation warning since IPython 8.0, it had no effects for a long time and will  be removed in future versions.\n",
      "  displaypub.publish_display_data(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Attaching SeuratObject\n",
       "In addition: Warning message:\n",
       "In (function (package, help, pos = 2, lib.loc = NULL, character.only = FALSE,  :\n",
       "  library ‘/usr/lib/R/site-library’ contains no packages\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "# R packages\n",
    "library(Seurat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-09-10 10:56:01--  https://figshare.com/ndownloader/files/45452260\n",
      "Resolving figshare.com (figshare.com)... 34.242.30.50, 18.203.71.222, 2a05:d018:1f4:d003:7a90:7623:e6f9:1f98, ...\n",
      "Connecting to figshare.com (figshare.com)|34.242.30.50|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/45452260/openproblems_bmmc_multiome_genes_filtered.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20240910/eu-west-1/s3/aws4_request&X-Amz-Date=20240910T105602Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=910bbdeb1f2742a4463269f7a87dbcead0ce51cd5f848f6f6f50a2c07c888b8f [following]\n",
      "--2024-09-10 10:56:02--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/45452260/openproblems_bmmc_multiome_genes_filtered.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20240910/eu-west-1/s3/aws4_request&X-Amz-Date=20240910T105602Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=910bbdeb1f2742a4463269f7a87dbcead0ce51cd5f848f6f6f50a2c07c888b8f\n",
      "Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.57.147, 52.218.116.48, 52.92.2.80, ...\n",
      "Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.57.147|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 3116868580 (2.9G) [application/octet-stream]\n",
      "Saving to: ‘openproblems_bmmc_multiome_genes_filtered.h5ad’\n",
      "\n",
      "openproblems_bmmc_m 100%[===================>]   2.90G  43.4MB/s    in 70s     \n",
      "\n",
      "2024-09-10 10:57:12 (42.3 MB/s) - ‘openproblems_bmmc_multiome_genes_filtered.h5ad’ saved [3116868580/3116868580]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://figshare.com/ndownloader/files/45452260 -O openproblems_bmmc_multiome_genes_filtered.h5ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/specs/registry.py:275: OldFormatWarning: Element '/layers' was written without encoding metadata.\n",
      "  return self.callback(read_func, elem.name, elem, iospec=iospec)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/GEX_pct_counts_mt' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/GEX_n_counts' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/GEX_n_genes' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/GEX_size_factors' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/GEX_phase' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/GEX_phase' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/ATAC_nCount_peaks' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/ATAC_atac_fragments' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/ATAC_reads_in_peaks_frac' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/ATAC_blacklist_fraction' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/ATAC_nucleosome_signal' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/cell_type' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/cell_type' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/batch' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/batch' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/ATAC_pseudotime_order' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/GEX_pseudotime_order' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/Samplename' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/Samplename' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/Site' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/Site' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/DonorNumber' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/DonorNumber' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/Modality' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/Modality' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/VendorLot' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/DonorID' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/DonorAge' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/DonorBMI' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/DonorBloodType' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/DonorBloodType' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/DonorRace' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/DonorRace' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/Ethnicity' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/Ethnicity' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/DonorGender' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/DonorGender' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/QCMeds' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/QCMeds' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/__categories/DonorSmoker' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/DonorSmoker' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/obs/_index' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/specs/registry.py:275: OldFormatWarning: Element '/obsm' was written without encoding metadata.\n",
      "  return self.callback(read_func, elem.name, elem, iospec=iospec)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/specs/registry.py:275: OldFormatWarning: Element '/obsm/ATAC_lsi_full' was written without encoding metadata.\n",
      "  return self.callback(read_func, elem.name, elem, iospec=iospec)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/specs/registry.py:275: OldFormatWarning: Element '/obsm/ATAC_lsi_red' was written without encoding metadata.\n",
      "  return self.callback(read_func, elem.name, elem, iospec=iospec)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/specs/registry.py:275: OldFormatWarning: Element '/obsm/ATAC_umap' was written without encoding metadata.\n",
      "  return self.callback(read_func, elem.name, elem, iospec=iospec)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/specs/registry.py:275: OldFormatWarning: Element '/obsm/GEX_X_pca' was written without encoding metadata.\n",
      "  return self.callback(read_func, elem.name, elem, iospec=iospec)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/specs/registry.py:275: OldFormatWarning: Element '/obsm/GEX_X_umap' was written without encoding metadata.\n",
      "  return self.callback(read_func, elem.name, elem, iospec=iospec)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/specs/registry.py:275: OldFormatWarning: Element '/uns' was written without encoding metadata.\n",
      "  return self.callback(read_func, elem.name, elem, iospec=iospec)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/specs/registry.py:275: OldFormatWarning: Element '/uns/ATAC_gene_activity_var_names' was written without encoding metadata.\n",
      "  return self.callback(read_func, elem.name, elem, iospec=iospec)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/specs/registry.py:275: OldFormatWarning: Element '/uns/dataset_id' was written without encoding metadata.\n",
      "  return self.callback(read_func, elem.name, elem, iospec=iospec)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/specs/registry.py:275: OldFormatWarning: Element '/uns/genome' was written without encoding metadata.\n",
      "  return self.callback(read_func, elem.name, elem, iospec=iospec)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/specs/registry.py:275: OldFormatWarning: Element '/uns/organism' was written without encoding metadata.\n",
      "  return self.callback(read_func, elem.name, elem, iospec=iospec)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/var/__categories/feature_types' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/var/feature_types' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/var/__categories/gene_id' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/var/gene_id' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_io/utils.py:211: OldFormatWarning: Element '/var/_index' was written without encoding metadata.\n",
      "  return func(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "adata_raw = sc.read_h5ad(\"openproblems_bmmc_multiome_genes_filtered.h5ad\")\n",
    "adata_raw.layers[\"logcounts\"] = adata_raw.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "label_key = \"cell_type\"\n",
    "batch_key = \"batch\"\n",
    "\n",
    "# adata_raw.obs[batch_key].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_batches = ['s3d10', 's1d2', 's2d1'] # S1, S2, S3 represent different facilitites\n",
    "adata          = adata_raw[adata_raw.obs[batch_key].isin(subset_batches)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Choose Gene Expression and Require Min Cell Threshold for Genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression = 'GEX'\n",
    "atac_seq        = 'ATAC'\n",
    "min_cells       = 1\n",
    "\n",
    "adata           = adata[:, adata.var['feature_types'] == gene_expression].copy()\n",
    "sc.pp.filter_genes(adata, min_cells=min_cells)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Choose our own normalization scheme from Raw Counts**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers['counts'].copy()\n",
    "\n",
    "sc.pp.normalize_total(adata)\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "# Reset Log Coutns variable\n",
    "adata.layers['logcounts'] = adata.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(adata)\n",
    "sc.tl.pca(adata)\n",
    "sc.pp.neighbors(adata)\n",
    "sc.tl.umap(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns[batch_key + \"_colors\"] = [\n",
    "    \"#1b9e77\",\n",
    "    \"#d95f02\",\n",
    "    \"#7570b3\",\n",
    "]  # Set custom colours for batches\n",
    "sc.pl.umap(adata, color=[label_key, batch_key], wspace=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(\n",
    "    adata, n_top_genes=2000, flavor=\"cell_ranger\", batch_key=batch_key\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "highly_variable_nbatches\n",
       "0    10239\n",
       "1     1397\n",
       "3     1013\n",
       "2      782\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var[\"highly_variable_nbatches\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 17741 × 13431\n",
       "    obs: 'GEX_pct_counts_mt', 'GEX_n_counts', 'GEX_n_genes', 'GEX_size_factors', 'GEX_phase', 'ATAC_nCount_peaks', 'ATAC_atac_fragments', 'ATAC_reads_in_peaks_frac', 'ATAC_blacklist_fraction', 'ATAC_nucleosome_signal', 'cell_type', 'batch', 'ATAC_pseudotime_order', 'GEX_pseudotime_order', 'Samplename', 'Site', 'DonorNumber', 'Modality', 'VendorLot', 'DonorID', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'QCMeds', 'DonorSmoker'\n",
       "    var: 'feature_types', 'gene_id', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection'\n",
       "    uns: 'ATAC_gene_activity_var_names', 'dataset_id', 'genome', 'organism', 'log1p', 'hvg', 'pca', 'neighbors', 'umap'\n",
       "    obsm: 'ATAC_gene_activity', 'ATAC_lsi_full', 'ATAC_lsi_red', 'ATAC_umap', 'GEX_X_pca', 'GEX_X_umap', 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'logcounts'\n",
       "    obsp: 'distances', 'connectivities'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Save Highly Variable Genes in Separate AnnData Object**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_hvg   = adata[:, adata.var['highly_variable']].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Single Cell Variational AutoEncoder INference (scVI)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting scvi-tools\n",
      "  Downloading scvi_tools-1.1.6-py3-none-any.whl.metadata (17 kB)\n",
      "Requirement already satisfied: anndata>=0.7.5 in /opt/conda/lib/python3.10/site-packages (from scvi-tools) (0.10.9)\n",
      "Collecting docrep>=0.3.2 (from scvi-tools)\n",
      "  Downloading docrep-0.3.2.tar.gz (33 kB)\n",
      "  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hCollecting flax (from scvi-tools)\n",
      "  Downloading flax-0.9.0-py3-none-any.whl.metadata (11 kB)\n",
      "Requirement already satisfied: h5py>=2.9.0 in /opt/conda/lib/python3.10/site-packages (from scvi-tools) (3.11.0)\n",
      "Collecting jax>=0.4.4 (from scvi-tools)\n",
      "  Downloading jax-0.4.31-py3-none-any.whl.metadata (22 kB)\n",
      "Collecting jaxlib>=0.4.3 (from scvi-tools)\n",
      "  Downloading jaxlib-0.4.31-cp310-cp310-manylinux2014_x86_64.whl.metadata (983 bytes)\n",
      "Collecting lightning<2.2,>=2.0 (from scvi-tools)\n",
      "  Downloading lightning-2.1.4-py3-none-any.whl.metadata (57 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m57.2/57.2 kB\u001b[0m \u001b[31m6.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hCollecting ml-collections>=0.1.1 (from scvi-tools)\n",
      "  Downloading ml_collections-0.1.1.tar.gz (77 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m77.9/77.9 kB\u001b[0m \u001b[31m8.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hCollecting mudata>=0.1.2 (from scvi-tools)\n",
      "  Downloading mudata-0.3.1-py3-none-any.whl.metadata (8.3 kB)\n",
      "Requirement already satisfied: numpy<2.0,>=1.21.0 in /opt/conda/lib/python3.10/site-packages (from scvi-tools) (1.26.4)\n",
      "Collecting numpyro>=0.12.1 (from scvi-tools)\n",
      "  Downloading numpyro-0.15.2-py3-none-any.whl.metadata (36 kB)\n",
      "Collecting optax (from scvi-tools)\n",
      "  Downloading optax-0.2.3-py3-none-any.whl.metadata (8.3 kB)\n",
      "Requirement already satisfied: pandas>=1.0 in /opt/conda/lib/python3.10/site-packages (from scvi-tools) (2.1.4)\n",
      "Collecting pyro-ppl>=1.6.0 (from scvi-tools)\n",
      "  Downloading pyro_ppl-1.9.1-py3-none-any.whl.metadata (7.8 kB)\n",
      "Requirement already satisfied: rich>=12.0.0 in /opt/conda/lib/python3.10/site-packages (from scvi-tools) (13.7.1)\n",
      "Requirement already satisfied: scikit-learn>=0.21.2 in /opt/conda/lib/python3.10/site-packages (from scvi-tools) (1.4.2)\n",
      "Requirement already satisfied: scipy in /opt/conda/lib/python3.10/site-packages (from scvi-tools) (1.11.4)\n",
      "Requirement already satisfied: torch>=1.8.0 in /opt/conda/lib/python3.10/site-packages (from scvi-tools) (2.0.0.post104)\n",
      "Requirement already satisfied: torchmetrics>=0.11.0 in /opt/conda/lib/python3.10/site-packages (from scvi-tools) (1.0.3)\n",
      "Requirement already satisfied: tqdm>=4.56.0 in /opt/conda/lib/python3.10/site-packages (from scvi-tools) (4.66.5)\n",
      "Requirement already satisfied: array-api-compat!=1.5,>1.4 in /opt/conda/lib/python3.10/site-packages (from anndata>=0.7.5->scvi-tools) (1.8)\n",
      "Requirement already satisfied: exceptiongroup in /opt/conda/lib/python3.10/site-packages (from anndata>=0.7.5->scvi-tools) (1.2.2)\n",
      "Requirement already satisfied: natsort in /opt/conda/lib/python3.10/site-packages (from anndata>=0.7.5->scvi-tools) (8.4.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /opt/conda/lib/python3.10/site-packages (from anndata>=0.7.5->scvi-tools) (24.1)\n",
      "Requirement already satisfied: six in /opt/conda/lib/python3.10/site-packages (from docrep>=0.3.2->scvi-tools) (1.16.0)\n",
      "Requirement already satisfied: ml-dtypes>=0.2.0 in /opt/conda/lib/python3.10/site-packages (from jax>=0.4.4->scvi-tools) (0.2.0)\n",
      "Requirement already satisfied: opt-einsum in /opt/conda/lib/python3.10/site-packages (from jax>=0.4.4->scvi-tools) (3.3.0)\n",
      "Requirement already satisfied: PyYAML<8.0,>=5.4 in /opt/conda/lib/python3.10/site-packages (from lightning<2.2,>=2.0->scvi-tools) (6.0.1)\n",
      "Requirement already satisfied: fsspec<2025.0,>=2022.5.0 in /opt/conda/lib/python3.10/site-packages (from fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (2023.6.0)\n",
      "Requirement already satisfied: lightning-utilities<2.0,>=0.8.0 in /opt/conda/lib/python3.10/site-packages (from lightning<2.2,>=2.0->scvi-tools) (0.11.6)\n",
      "Requirement already satisfied: typing-extensions<6.0,>=4.0.0 in /opt/conda/lib/python3.10/site-packages (from lightning<2.2,>=2.0->scvi-tools) (4.12.2)\n",
      "Requirement already satisfied: pytorch-lightning in /opt/conda/lib/python3.10/site-packages (from lightning<2.2,>=2.0->scvi-tools) (2.0.9)\n",
      "Requirement already satisfied: absl-py in /opt/conda/lib/python3.10/site-packages (from ml-collections>=0.1.1->scvi-tools) (2.1.0)\n",
      "Requirement already satisfied: contextlib2 in /opt/conda/lib/python3.10/site-packages (from ml-collections>=0.1.1->scvi-tools) (21.6.0)\n",
      "Collecting multipledispatch (from numpyro>=0.12.1->scvi-tools)\n",
      "  Downloading multipledispatch-1.0.0-py3-none-any.whl.metadata (3.8 kB)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /opt/conda/lib/python3.10/site-packages (from pandas>=1.0->scvi-tools) (2.9.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /opt/conda/lib/python3.10/site-packages (from pandas>=1.0->scvi-tools) (2023.3)\n",
      "Requirement already satisfied: tzdata>=2022.1 in /opt/conda/lib/python3.10/site-packages (from pandas>=1.0->scvi-tools) (2024.1)\n",
      "Collecting pyro-api>=0.1.1 (from pyro-ppl>=1.6.0->scvi-tools)\n",
      "  Downloading pyro_api-0.1.2-py3-none-any.whl.metadata (2.5 kB)\n",
      "Requirement already satisfied: markdown-it-py>=2.2.0 in /opt/conda/lib/python3.10/site-packages (from rich>=12.0.0->scvi-tools) (3.0.0)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /opt/conda/lib/python3.10/site-packages (from rich>=12.0.0->scvi-tools) (2.18.0)\n",
      "Requirement already satisfied: joblib>=1.2.0 in /opt/conda/lib/python3.10/site-packages (from scikit-learn>=0.21.2->scvi-tools) (1.4.2)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /opt/conda/lib/python3.10/site-packages (from scikit-learn>=0.21.2->scvi-tools) (3.5.0)\n",
      "Requirement already satisfied: filelock in /opt/conda/lib/python3.10/site-packages (from torch>=1.8.0->scvi-tools) (3.15.4)\n",
      "Requirement already satisfied: sympy in /opt/conda/lib/python3.10/site-packages (from torch>=1.8.0->scvi-tools) (1.13.0)\n",
      "Requirement already satisfied: networkx in /opt/conda/lib/python3.10/site-packages (from torch>=1.8.0->scvi-tools) (3.3)\n",
      "Requirement already satisfied: jinja2 in /opt/conda/lib/python3.10/site-packages (from torch>=1.8.0->scvi-tools) (3.1.4)\n",
      "Requirement already satisfied: msgpack in /opt/conda/lib/python3.10/site-packages (from flax->scvi-tools) (1.0.8)\n",
      "Collecting orbax-checkpoint (from flax->scvi-tools)\n",
      "  Downloading orbax_checkpoint-0.6.1-py3-none-any.whl.metadata (1.8 kB)\n",
      "Collecting tensorstore (from flax->scvi-tools)\n",
      "  Downloading tensorstore-0.1.65-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.0 kB)\n",
      "Collecting chex>=0.1.86 (from optax->scvi-tools)\n",
      "  Downloading chex-0.1.86-py3-none-any.whl.metadata (17 kB)\n",
      "Collecting etils[epy] (from optax->scvi-tools)\n",
      "  Downloading etils-1.9.4-py3-none-any.whl.metadata (6.4 kB)\n",
      "Requirement already satisfied: toolz>=0.9.0 in /opt/conda/lib/python3.10/site-packages (from chex>=0.1.86->optax->scvi-tools) (0.12.1)\n",
      "Requirement already satisfied: requests in /opt/conda/lib/python3.10/site-packages (from fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (2.32.3)\n",
      "Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in /opt/conda/lib/python3.10/site-packages (from fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (3.9.5)\n",
      "Requirement already satisfied: setuptools in /opt/conda/lib/python3.10/site-packages (from lightning-utilities<2.0,>=0.8.0->lightning<2.2,>=2.0->scvi-tools) (67.0.0)\n",
      "Requirement already satisfied: mdurl~=0.1 in /opt/conda/lib/python3.10/site-packages (from markdown-it-py>=2.2.0->rich>=12.0.0->scvi-tools) (0.1.2)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /opt/conda/lib/python3.10/site-packages (from jinja2->torch>=1.8.0->scvi-tools) (2.1.5)\n",
      "Requirement already satisfied: nest_asyncio in /opt/conda/lib/python3.10/site-packages (from orbax-checkpoint->flax->scvi-tools) (1.6.0)\n",
      "Requirement already satisfied: protobuf in /opt/conda/lib/python3.10/site-packages (from orbax-checkpoint->flax->scvi-tools) (4.24.4)\n",
      "Collecting humanize (from orbax-checkpoint->flax->scvi-tools)\n",
      "  Downloading humanize-4.10.0-py3-none-any.whl.metadata (7.9 kB)\n",
      "Collecting ml-dtypes>=0.2.0 (from jax>=0.4.4->scvi-tools)\n",
      "  Downloading ml_dtypes-0.4.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (20 kB)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /opt/conda/lib/python3.10/site-packages (from sympy->torch>=1.8.0->scvi-tools) (1.3.0)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /opt/conda/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (1.3.1)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /opt/conda/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (23.2.0)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /opt/conda/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (1.4.1)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /opt/conda/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (6.0.5)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /opt/conda/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (1.9.4)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0 in /opt/conda/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (4.0.3)\n",
      "Requirement already satisfied: importlib_resources in /opt/conda/lib/python3.10/site-packages (from etils[epy]->optax->scvi-tools) (6.4.0)\n",
      "Requirement already satisfied: zipp in /opt/conda/lib/python3.10/site-packages (from etils[epy]->optax->scvi-tools) (3.19.2)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /opt/conda/lib/python3.10/site-packages (from requests->fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.10/site-packages (from requests->fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (3.7)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /opt/conda/lib/python3.10/site-packages (from requests->fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (1.26.19)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.10/site-packages (from requests->fsspec[http]<2025.0,>=2022.5.0->lightning<2.2,>=2.0->scvi-tools) (2024.7.4)\n",
      "Downloading scvi_tools-1.1.6-py3-none-any.whl (387 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m387.7/387.7 kB\u001b[0m \u001b[31m44.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading jax-0.4.31-py3-none-any.whl (2.0 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.0/2.0 MB\u001b[0m \u001b[31m89.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading jaxlib-0.4.31-cp310-cp310-manylinux2014_x86_64.whl (88.1 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m88.1/88.1 MB\u001b[0m \u001b[31m31.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading lightning-2.1.4-py3-none-any.whl (2.0 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.0/2.0 MB\u001b[0m \u001b[31m92.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading mudata-0.3.1-py3-none-any.whl (41 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m41.0/41.0 kB\u001b[0m \u001b[31m4.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading numpyro-0.15.2-py3-none-any.whl (348 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m348.1/348.1 kB\u001b[0m \u001b[31m42.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading pyro_ppl-1.9.1-py3-none-any.whl (755 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m756.0/756.0 kB\u001b[0m \u001b[31m61.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading flax-0.9.0-py3-none-any.whl (780 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m780.7/780.7 kB\u001b[0m \u001b[31m54.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading optax-0.2.3-py3-none-any.whl (289 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m289.6/289.6 kB\u001b[0m \u001b[31m33.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading chex-0.1.86-py3-none-any.whl (98 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m98.2/98.2 kB\u001b[0m \u001b[31m15.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading pyro_api-0.1.2-py3-none-any.whl (11 kB)\n",
      "Downloading multipledispatch-1.0.0-py3-none-any.whl (12 kB)\n",
      "Downloading orbax_checkpoint-0.6.1-py3-none-any.whl (243 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m243.8/243.8 kB\u001b[0m \u001b[31m24.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading tensorstore-0.1.65-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (14.9 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m14.9/14.9 MB\u001b[0m \u001b[31m93.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading ml_dtypes-0.4.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.2 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.2/2.2 MB\u001b[0m \u001b[31m92.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading etils-1.9.4-py3-none-any.whl (164 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m164.3/164.3 kB\u001b[0m \u001b[31m22.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading humanize-4.10.0-py3-none-any.whl (126 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m127.0/127.0 kB\u001b[0m \u001b[31m18.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hBuilding wheels for collected packages: docrep, ml-collections\n",
      "  Building wheel for docrep (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for docrep: filename=docrep-0.3.2-py3-none-any.whl size=19878 sha256=79ad8a92efea37e765f842d6e20ee571b2951225aff27d798fe5da8c50107b89\n",
      "  Stored in directory: /home/sagemaker-user/.cache/pip/wheels/c3/64/48/03c38d8d906159eaa210b3c548fdb590eb3e2a4a5745ae2172\n",
      "  Building wheel for ml-collections (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for ml-collections: filename=ml_collections-0.1.1-py3-none-any.whl size=94505 sha256=2db46c7feee0b4e07de0edb4e15aa0fd96b8bebcb3906bc6b2105a476041e9ad\n",
      "  Stored in directory: /home/sagemaker-user/.cache/pip/wheels/7b/89/c9/a9b87790789e94aadcfc393c283e3ecd5ab916aed0a31be8fe\n",
      "Successfully built docrep ml-collections\n",
      "Installing collected packages: pyro-api, multipledispatch, ml-dtypes, ml-collections, humanize, etils, docrep, tensorstore, jaxlib, pyro-ppl, jax, orbax-checkpoint, numpyro, mudata, chex, optax, lightning, flax, scvi-tools\n",
      "  Attempting uninstall: ml-dtypes\n",
      "    Found existing installation: ml-dtypes 0.2.0\n",
      "    Uninstalling ml-dtypes-0.2.0:\n",
      "      Successfully uninstalled ml-dtypes-0.2.0\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "tensorflow 2.15.0 requires ml-dtypes~=0.2.0, but you have ml-dtypes 0.4.0 which is incompatible.\u001b[0m\u001b[31m\n",
      "\u001b[0mSuccessfully installed chex-0.1.86 docrep-0.3.2 etils-1.9.4 flax-0.9.0 humanize-4.10.0 jax-0.4.31 jaxlib-0.4.31 lightning-2.1.4 ml-collections-0.1.1 ml-dtypes-0.4.0 mudata-0.3.1 multipledispatch-1.0.0 numpyro-0.15.2 optax-0.2.3 orbax-checkpoint-0.6.1 pyro-api-0.1.2 pyro-ppl-1.9.1 scvi-tools-1.1.6 tensorstore-0.1.65\n"
     ]
    }
   ],
   "source": [
    "!pip install scvi-tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'scvi' has no attribute 'model'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m/home/sagemaker-user/sc-RNAseq-embryos/Batch_effects.ipynb Cell 21\u001b[0m line \u001b[0;36m4\n\u001b[1;32m      <a href='vscode-notebook-cell://yyufmplsdbk3gbo.studio.us-east-1.sagemaker.aws/home/sagemaker-user/sc-RNAseq-embryos/Batch_effects.ipynb#X30sdnNjb2RlLXJlbW90ZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mscvi\u001b[39;00m \n\u001b[1;32m      <a href='vscode-notebook-cell://yyufmplsdbk3gbo.studio.us-east-1.sagemaker.aws/home/sagemaker-user/sc-RNAseq-embryos/Batch_effects.ipynb#X30sdnNjb2RlLXJlbW90ZQ%3D%3D?line=2'>3</a>\u001b[0m adata_scvi \u001b[39m=\u001b[39m adata_hvg\u001b[39m.\u001b[39mcopy()\n\u001b[0;32m----> <a href='vscode-notebook-cell://yyufmplsdbk3gbo.studio.us-east-1.sagemaker.aws/home/sagemaker-user/sc-RNAseq-embryos/Batch_effects.ipynb#X30sdnNjb2RlLXJlbW90ZQ%3D%3D?line=3'>4</a>\u001b[0m scvi\u001b[39m.\u001b[39;49mmodel\u001b[39m.\u001b[39mSCVI\u001b[39m.\u001b[39msetup_anndata(adata_scvi, layer\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mcounts\u001b[39m\u001b[39m\"\u001b[39m, batch_key\u001b[39m=\u001b[39mbatch_key)\n",
      "\u001b[0;31mAttributeError\u001b[0m: module 'scvi' has no attribute 'model'"
     ]
    }
   ],
   "source": [
    "import scvi \n",
    "\n",
    "adata_scvi = adata_hvg.copy()\n",
    "scvi.model.SCVI.setup_anndata(adata_scvi, layer=\"counts\", batch_key=batch_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package scvi:\n",
      "\n",
      "NAME\n",
      "    scvi - Top-level package for scVI-dev.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    _settings\n",
      "    benchmark\n",
      "    dataset (package)\n",
      "    inference (package)\n",
      "    models (package)\n",
      "\n",
      "FUNCTIONS\n",
      "    set_seed(seed: int = 0)\n",
      "    \n",
      "    set_verbosity(level: Union[str, int])\n",
      "        Sets logging configuration for scvi based on chosen level of verbosity.\n",
      "        \n",
      "        Sets \"scvi\" logging level to `level`\n",
      "        If \"scvi\" logger has no StreamHandler, add one.\n",
      "        Else, set its level to `level`.\n",
      "\n",
      "DATA\n",
      "    __all__ = ['set_verbosity', 'set_seed']\n",
      "    __email__ = 'romain_lopez@berkeley.edu'\n",
      "    __warningregistry__ = {'version': 149}\n",
      "\n",
      "VERSION\n",
      "    0.6.8\n",
      "\n",
      "AUTHOR\n",
      "    Romain Lopez\n",
      "\n",
      "FILE\n",
      "    /opt/conda/lib/python3.10/site-packages/scvi/__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(scvi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'scvi' has no attribute 'model'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m/home/sagemaker-user/sc-RNAseq-embryos/Batch_effects.ipynb Cell 23\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell://yyufmplsdbk3gbo.studio.us-east-1.sagemaker.aws/home/sagemaker-user/sc-RNAseq-embryos/Batch_effects.ipynb#X32sdnNjb2RlLXJlbW90ZQ%3D%3D?line=0'>1</a>\u001b[0m model_scvi \u001b[39m=\u001b[39m scvi\u001b[39m.\u001b[39;49mmodel\u001b[39m.\u001b[39mSCVI(adata_scvi)\n",
      "\u001b[0;31mAttributeError\u001b[0m: module 'scvi' has no attribute 'model'"
     ]
    }
   ],
   "source": [
    "model_scvi = scvi.model.SCVI(adata_scvi)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
